docs/content/contrib-adding-builtin-functions.md

```diff
@@ -0,0 +1,156 @@
+---
+title: Adding Built-in Functions
+kind: contrib
+weight: 5
+---
+
+[Built-in Functions](../policy-reference/#built-in-functions)
+can be added inside the `topdown` package.
+
+Built-in functions may be upstreamed if they are generally useful and provide functionality that would be
+impractical to implement natively in Rego (e.g., CIDR arithmetic). Implementations should avoid thirdparty
+dependencies. If absolutely necessary, consider importing the code manually into the `internal` package.
+
+Adding a new built-in function involves the following steps:
+
+1. [Declare and register](#declare-and-register) the function
+2. [Implementation](#implement) the function
+3. [Test](#test) the function
+4. [Document](#document) the function
+
+## Example
+
+The following example adds a simple built-in function, `repeat(string, int)`, that returns a given string repeated a given number of times.  
+
+### Declare and Register
+
+In `ast/builtins.go`, we declare the structure of our built-in function with a `Builtin` struct instance:
+
+```go
+// Repeat returns, as a string, the given string repeated the given number of times.
+var Repeat = &Builtin{
+    Name: "repeat",  // The name of the function
+    Decl: types.NewFunction(
+        types.Args(  // The built-in takes two arguments, where ..
+            types.S, // .. the first is a string, and ..
+            types.N, // .. the second is a number.
+        ),
+        types.S, // The return type is a string.
+    ),
+}
+```
+
+To register the new built-in function, we locate the `DefaultBuiltins` array in `ast/builtins.go`, and add the `Builtin` instance to it:
+
+```go
+var DefaultBuiltins = [...]*Builtin{
+    ...
+    Repeat,
+    ...
+}
+```
+
+### Implement
+
+In the `topdown` package, we locate a suitable source file for our new built-in function, or add a new file, as appropriate.
+
+In this example, we introduce a new source file, `topdown/repeat.go`:
+
+```go
+package topdown
+
+import (
+    "github.com/open-policy-agent/opa/ast"
+    "github.com/open-policy-agent/opa/topdown/builtins"
+    "github.com/pkg/errors"
+    "strings"
+)
```

Comment:
[nit] Let's make it three blocks: stdlib, 3rd party, internal:
```suggestion
import (
    "strings"
    
    "github.com/pkg/errors"
    
    "github.com/open-policy-agent/opa/ast"
    "github.com/open-policy-agent/opa/topdown/builtins"
)
```

and let's not provide examples using pkg/errors: https://github.com/open-policy-agent/opa/issues/2152